#Load and normalize the Fashion MNIST dataset

Like the previous lab, you will use the fashion MNIST dataset again for this exercise. And also as mentioned before, you will normalize the pixel values to help optimize the training.


In [2]:
import tensorflow as tf

# Instantizte the dataset API
fmnist = tf.keras.datasets.fashion_mnist

# load the dataset
(x_train, y_train), (x_test, y_test) = fmnist.load_data()

# Normalize the pixel values
x_train, x_test = x_train / 255.0, x_test / 255.0

# Creating callback class

You can create a callback by defining a class that inherits the tf.keras.callbacks.Callback base class. From there, you can define available methods to set where the callback will be executed. For instance below, you will use the on_epoch_end() method to check the loss at training epoch

In [5]:
class myCallback(tf.keras.callbacks.Callback):
  def on_epoch_end(self, epoch, logs={}):
    '''
    Halts the training after reaching 60 percent accuracy

    Args:
      epoch (integer) - index of epoch (required but unused in the function definition below)
      logs (dict) - metric results from the training epoch
    '''

    # Check accuracy
    if(logs.get('loss') < 0.4):

      # Stop if threshold is met
      print("\nLoss is lower than 0.4 so cancelling training!")
      self.model.stop_training = True

# Instantiate class
callbacks = myCallback()

# Define and compile the model

Next, you will define and compile the model. The architecture will be similar to the one you built in the previous lab. Afterwards, you will set the optimizer, loss, and metrics that you will use for training.

In [7]:
# Define the model
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape = (28, 28)),
    tf.keras.layers.Dense(512, activation = tf.nn.relu),
    tf.keras.layers.Dense(10, activation = tf.nn.softmax)
])

# Compile the model
model.compile(optimizer = tf.optimizers.Adam(), 
              loss = 'sparse_categorical_crossentropy',
              metrics = ['accuracy'])

# Train the model

Now you are ready to train the model. To set the callback, simply set the callbacks parameter to the myCallback instanyou declared before. Run the cell below and observe what happens.

In [9]:
# Train the model wit a callback
model.fit(x_train, y_train, epochs = 10, callbacks = [callbacks])

Epoch 1/10
1875/1875 [==============================] - 11s 5ms/step - loss: 0.4738 - accuracy: 0.8306
Epoch 2/10
1875/1875 [==============================] - ETA: 0s - loss: 0.3609 - accuracy: 0.8672
Loss is lower than 0.4 so cancelling training!
1875/1875 [==============================] - 9s 5ms/step - loss: 0.3609 - accuracy: 0.8672


You will notice that the training does not need to complete all 10 epochs. By having a callback at each end of the epoch. It is able to check the training parameters and compare if it meets the threshold you set in the function definition. In this case, it will simply stop when the loss falls below 0.40 after the current epoch.

